Crash trying to load tabPageIdentifiers() and popupPageIdentifiers().

xeenon opened the pull request:
#### 2f4ea4c094b72386cd25be293b6b6fa2c67064b1
<pre>
Crash trying to load tabPageIdentifiers() and popupPageIdentifiers().
<a href="https://webkit.org/b/268105">https://webkit.org/b/268105</a>
<a href="https://rdar.apple.com/120867910">rdar://120867910</a>

Reviewed by Brian Weinstein.

Clean up how we track extension tab and action pages, so we don&apos;t need to loop over openTabs(),
which is not safe to do while a WKWebView is being created.

Also clean up WKWebViewConfiguration, by not storing the WebExtensionController as
data members, when they are already stored on API::PageConfiguration. Also added
requiredWebExtensionBaseURL() to API::PageConfiguration.

Also made a few more tab and window getters on WebExtensionContext be const, to allow
use inside other const functions, like parameters().

* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::requiredWebExtensionBaseURL const): Added.
(API::PageConfiguration::setRequiredWebExtensionBaseURL): Added.
(API::PageConfiguration::webExtensionController const): Made const.
(API::PageConfiguration::weakWebExtensionController const): Ditto.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _requiredWebExtensionBaseURL]): Use APIPageConfiguration.
(-[WKWebViewConfiguration _setRequiredWebExtensionBaseURL:]): Ditto.
(-[WKWebViewConfiguration _strongWebExtensionController]): Ditto.
(-[WKWebViewConfiguration _webExtensionController]): Ditto.
(-[WKWebViewConfiguration _setWebExtensionController:]): Ditto.
(-[WKWebViewConfiguration _weakWebExtensionController]): Ditto.
(-[WKWebViewConfiguration _setWeakWebExtensionController:]): Ditto.
(-[WKWebViewConfiguration _setShouldRelaxThirdPartyCookieBlocking:]): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::fallbackAction const): Added helper.
(WebKit::WebExtensionAction::icon): Use fallbackAction().
(WebKit::WebExtensionAction::popupPath const): Ditto.
(WebKit::WebExtensionAction::popupWebView): Use new addPopupPage() method.
(WebKit::WebExtensionAction::label const): Use fallbackAction().
(WebKit::WebExtensionAction::badgeText const): Ditto.
(WebKit::WebExtensionAction::isEnabled const): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::isURLForThisExtension const): Added isValid() to bail earlier.
(WebKit::WebExtensionContext::getOrCreateWindow const): Made const.
(WebKit::WebExtensionContext::getWindow const): Ditto.
(WebKit::WebExtensionContext::getOrCreateTab const): Ditto.
(WebKit::WebExtensionContext::getTab const): Ditto.
(WebKit::WebExtensionContext::getCurrentTab const): Ditto.
(WebKit::WebExtensionContext::focusedWindow const): Ditto.
(WebKit::WebExtensionContext::frontmostWindow const): Ditto.
(WebKit::WebExtensionContext::popupPageIdentifiers const): Use the new map to build the Vector.
(WebKit::WebExtensionContext::tabPageIdentifiers const): Use the new map to build the Vector.
(WebKit::WebExtensionContext::addExtensionTabPage): Added.
(WebKit::WebExtensionContext::relatedWebView): Simplified by using requiredWebExtensionBaseURL().
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask): Call addExtensionTabPage().
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::parameters const): Use WTF::map().
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIExtension.mm:
(TestWebKitAPI::TEST): Tweak test to conver another case. Correct test that was logging a JSON
error for object cycles.

Canonical link: <a href="https://commits.webkit.org/273550@main">https://commits.webkit.org/273550@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/094b4dda3baf6aeb0e59d712f854857e0b93cab0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35779 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14722 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37925 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38504 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32196 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36997 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17116 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11746 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36333 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12426 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31807 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10926 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/10918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31976 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39751 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32492 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32300 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36874 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/11114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/9005 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/34972 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12853 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/11619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4640 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11931 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2375,15 +2376,26 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
-        auto window = tab->window();
-        for (WKWebView *webView in tab->webViews())
-            result.append({ webView._page->webPageID(), tab->identifier(), window ? std::optional(window->identifier()) : std::nullopt });
+    for (auto entry : m_extensionPageTabMap) {
```
should `WebExtensionContext::getTab` also iterate through `m_extensionPageTabMap` instead of `openTabs())`?

xeenon replied:
Yeah I think that would be good!

b-weinstein reviewed (commented):
`by no storing the WebExtensionController`

by not storing the...

xeenon committed:
Crash trying to load tabPageIdentifiers() and popupPageIdentifiers().
https://webkit.org/b/268105
rdar://120867910

Reviewed by Brian Weinstein.

Clean up how we track extension tab and action pages, so we don't need to loop over openTabs(),
which is not safe to do while a WKWebView is being created.

Also clean up WKWebViewConfiguration, by not storing the WebExtensionController as
data members, when they are already stored on API::PageConfiguration. Also added
requiredWebExtensionBaseURL() to API::PageConfiguration.

Also made a few more tab and window getters on WebExtensionContext be const, to allow
use inside other const functions, like parameters().

* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::requiredWebExtensionBaseURL const): Added.
(API::PageConfiguration::setRequiredWebExtensionBaseURL): Added.
(API::PageConfiguration::webExtensionController const): Made const.
(API::PageConfiguration::weakWebExtensionController const): Ditto.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration copyWithZone:]):
(-[WKWebViewConfiguration _requiredWebExtensionBaseURL]): Use APIPageConfiguration.
(-[WKWebViewConfiguration _setRequiredWebExtensionBaseURL:]): Ditto.
(-[WKWebViewConfiguration _strongWebExtensionController]): Ditto.
(-[WKWebViewConfiguration _webExtensionController]): Ditto.
(-[WKWebViewConfiguration _setWebExtensionController:]): Ditto.
(-[WKWebViewConfiguration _weakWebExtensionController]): Ditto.
(-[WKWebViewConfiguration _setWeakWebExtensionController:]): Ditto.
(-[WKWebViewConfiguration _setShouldRelaxThirdPartyCookieBlocking:]): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::fallbackAction const): Added helper.
(WebKit::WebExtensionAction::icon): Use fallbackAction().
(WebKit::WebExtensionAction::popupPath const): Ditto.
(WebKit::WebExtensionAction::popupWebView): Use new addPopupPage() method.
(WebKit::WebExtensionAction::label const): Use fallbackAction().
(WebKit::WebExtensionAction::badgeText const): Ditto.
(WebKit::WebExtensionAction::isEnabled const): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::isURLForThisExtension const): Added isValid() to bail earlier.
(WebKit::WebExtensionContext::getOrCreateWindow const): Made const.
(WebKit::WebExtensionContext::getWindow const): Ditto.
(WebKit::WebExtensionContext::getOrCreateTab const): Ditto.
(WebKit::WebExtensionContext::getTab const): Ditto.
(WebKit::WebExtensionContext::getCurrentTab const): Ditto.
(WebKit::WebExtensionContext::focusedWindow const): Ditto.
(WebKit::WebExtensionContext::frontmostWindow const): Ditto.
(WebKit::WebExtensionContext::popupPageIdentifiers const): Use the new map to build the Vector.
(WebKit::WebExtensionContext::tabPageIdentifiers const): Use the new map to build the Vector.
(WebKit::WebExtensionContext::addExtensionTabPage): Added.
(WebKit::WebExtensionContext::relatedWebView): Simplified by using requiredWebExtensionBaseURL().
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask): Call addExtensionTabPage().
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::parameters const): Use WTF::map().
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIExtension.mm:
(TestWebKitAPI::TEST): Tweak test to conver another case. Correct test that was logging a JSON
error for object cycles.

Canonical link: https://commits.webkit.org/273550@main
